aiida/manage/database/integrity.py

```diff
@@ -0,0 +1,69 @@
+# -*- coding: utf-8 -*-
+"""Generic functions to verify the integrity of the database and optionally apply patches to fix problems."""
+from __future__ import absolute_import
+import uuid as UUID
+
+
+def deduplicate_node_uuids(dry_run=True):
+    """Detect and solve nodes with duplicate UUIDs.
+
+    Before aiida-core v1.0.0, there was no uniqueness constraint on the UUID column of the Node table in the database.
+    This made it possible to store multiple nodes with identical UUIDs without the database complaining. This was bug
+    was fixed in aiida-core=1.0.0 by putting an explicit uniqueness constraint on node UUIDs on the database level.
+    However, this would leave databases created before this patch with duplicate UUIDs in an inconsistent state. This
+    command will run an analysis to detect duplicate UUIDs in the node table and solve it by generating new UUIDs. Note
+    that it will not delete or merge any nodes.
+
+    :param dry_run: when True, no actual changes will be made
+    :return: list of strings denoting the performed operations, or those that would have been applied for dry_run=False
+    """
+    from aiida.backends.settings import AIIDANODES_UUID_VERSION
+    from aiida.orm import load_node
+    from aiida.orm.backend import construct_backend
+
+    uuid_generator = getattr(UUID, 'uuid{}'.format(AIIDANODES_UUID_VERSION))
+
+    backend = construct_backend()
+
+    query = 'SELECT s.id, s.uuid FROM (SELECT *, COUNT(*) OVER(PARTITION BY uuid) AS c FROM db_dbnode) AS s WHERE c > 1'
+    duplicates = backend.query_manager.raw(query)
+
+    mapping = {}
```

Comment:
I think it's easier/cleaner to use `mapping = defaultdict(list)` (it's In `from collections import defaultdict`) and then you can simply append, without setdefaullt.

Reply:
agreed

Reply:
done